File: resources/features/steps/duplicate_marker_steps.py
from behave import Given, When, Then
from python.sml.methods.duplicate import duplicate


@Given('the user provides the partition and order columns')
def step_impl(context):
    context.partition_cols = str(context.table.rows[0][0]).split(",")
    context.order_cols = str(context.table.rows[0][1]).split(",")


@When('the Python Duplicate Marker function is applied')
def step_impl(context):
    new_col = "Duplicate Marker"
    context.output_df = duplicate(context.input_data).dm1(context.input_data, context.partition_cols,
                                                        context.order_cols, new_col)
    print("Expected DataFrame ...")
    context.expected_data.select("id", "num", "order", "Duplicate Marker").show()
    print("Output DataFrame ...")
    context.output_df.select("id", "num", "order", "Duplicate Marker").show()


@Then('record with highest order column is marked with one to match expected data')
def step_impl(context):
    assert context.expected_data.select("id", "num", "order", "Duplicate Marker").collect() == \
                                              context.output_df.select("id", "num", "order",
                                                                       "Duplicate Marker").collect()

File: src/test/scala/uk/gov/ons/stepdefs/DuplicateMarkerSteps.scala
package uk.gov.ons.stepdefs
import uk.gov.ons.methods.Duplicate
import cucumber.api.scala.{EN, ScalaDsl}
import org.apache.spark.sql.DataFrame
import cucumber.api.DataTable
import scala.collection.JavaConversions._

object ContextDuplicate {

  var partition_cols : Seq[String] = _
  var order_cols : Seq[String] = _
  var input_data: DataFrame = _
  var expected_data: DataFrame = _
  var partition_list: List[String] = _
  var order_list: List[String] = _
}
//noinspection ScalaStyle
class DuplicateMarkerSteps extends ScalaDsl with EN {


  Given("""^the user provides the partition and order columns:$"""){ (x:DataTable) =>

    val inputList = x.asLists(classOf[String])
    ContextCommon.param_list = inputList.get(1).toSeq
    ContextDuplicate.input_data = ContextCommon.input_data
    ContextDuplicate.partition_cols = ContextCommon.param_list.head.split(",")
    ContextDuplicate.partition_list = ContextDuplicate.partition_cols.toList
    ContextDuplicate.order_cols = ContextCommon.param_list(1).split(",")
    ContextDuplicate.order_list = ContextDuplicate.order_cols.toList

  }

  When("""^the Scala Duplicate Marker function is applied$""") { () =>

    ContextCommon.output_data = Duplicate.duplicate(ContextDuplicate.input_data)
      .dm1(ContextCommon.input_data, ContextDuplicate.partition_list,
        ContextDuplicate.order_list, DuplicateContext.new_column_name)
    println("Scala Duplicate Marker DataFrame")
    println("Scala Duplicate DataFrame")
    ContextCommon.output_data.show()

  }

  Then("""^record with highest order column is marked with one to match expected data$""") { () =>

    println("Expected DataFrame")
    ContextCommon.expected_data.show()
    assert(ContextCommon.expected_data.select("id", "num", "order", "Duplicate Marker").collect()
      sameElements ContextCommon.output_data.select("id", "num", "order", "Duplicate Marker").collect())
  }

}

object DuplicateContext {
  var new_column_name: String = "Duplicate Marker"
}
